src/common/utils.h

```diff
@@ -43,9 +43,84 @@
 #include <algorithm>
 #include <functional>
 
+#include "../operator/mxnet_op.h"
+#include "../ndarray/ndarray_function.h"
+
 namespace mxnet {
 namespace common {
 
+struct indptr_check {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, mshadow::default_real_t* out, const DType* in,
+                                  const nnvm::dim_t end, const nnvm::dim_t idx_size) {
+    if ((in[i+1] < in[i]) || (i == 0 && in[i] != static_cast<DType>(0)) ||
+        (i == end && in[i] < static_cast<DType>(idx_size))) out[0] = 1;
+  }
+};
+
+struct idx_check {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, mshadow::default_real_t* out,
+                                  const DType* in, const nnvm::dim_t ncols) {
+    if (in[i] >= static_cast<DType>(ncols)) out[0] = 1;
+  }
+};
+
+template<typename xpu>
+void CheckFormatWrapper(const RunContext &rctx, const NDArray *input,
+                        NDArray *cpu_ret, const bool &full_check);
+
+template<typename xpu>
+void CheckFormatImpl(const RunContext &rctx, const NDArray *input,
+                     NDArray *cpu_ret, const bool &full_check) {
```

Comment:
Is it possible to pass `ret` on the same ctx as `input` and avoid all these `stream->wait` calls in this function?

Reply:
`xpu_ret ` is on the same context with `input`. `cpu_ret` here is used for inspecting data (err number) on cpu.

Reply:
`xpu_ret` is used for kernel launch with `input`, which can be on cpu/gpu. Then copy the `err` number to `cpu_ret` for checking outside of engine, mainly for inspecting data.
```
auto err = cpu_ret.data().dptr<mshadow::default_real_t>();
CHECK_EQ(*err, 0) << "Check validity of the CSRNDArray";
```
If data is on gpu, can't do like this.

Reply:
`stream->wait` is used for `Eval` and `Copy` with `input` and `xpu_ret`. I think these `wait` are neccessary for the corresponding computations.